### PR TITLE
added order

### PR DESCRIPTION
**Pull request by TarasYurkovskyi**

### CODE REVIEW

File: src/main/java/com/example/bookstore/controller/OrderController.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestController;
+
+@Tag(name = "Order management", description = "Endpoints for managing orders")
+@RestController
+@RequestMapping(value = "/api/orders")
```
Comment (BohdanTrue) [inline src/main/java/com/example/bookstore/controller/OrderController.java:30]: Not critical, but you can add this property `server.servlet.context-path=/api` to application.properties and write all RequestMapping's without prefix `/api`

File: src/main/java/com/example/bookstore/model/Order.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+@NoArgsConstructor
+@Data
+@Table(name = "orders")
+public class Order {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ToString.Exclude
+    @EqualsAndHashCode.Exclude
+    @ManyToOne
```
Comment (BohdanTrue) [inline src/main/java/com/example/bookstore/model/Order.java]: `@ManyToOne(fetch = FetchType.LAZY)`

File: src/main/java/com/example/bookstore/model/OrderItem.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+@NoArgsConstructor
+@Data
+@Table(name = "orders_items")
+public class OrderItem {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ToString.Exclude
+    @EqualsAndHashCode.Exclude
+    @ManyToOne
```
Comment (BohdanTrue) [inline src/main/java/com/example/bookstore/model/OrderItem.java]: `@ManyToOne(fetch = FetchType.LAZY)`

File: src/main/java/com/example/bookstore/model/OrderItem.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+import java.math.BigDecimal;
+import lombok.Data;
+import lombok.EqualsAndHashCode;
+import lombok.NoArgsConstructor;
+import lombok.ToString;
+
+@Entity
+@NoArgsConstructor
+@Data
+@Table(name = "orders_items")
```
Comment (BohdanTrue) [inline src/main/java/com/example/bookstore/model/OrderItem.java]: `@Table(name = "order_items")`

File: src/main/java/com/example/bookstore/model/OrderItem.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ToString.Exclude
+    @EqualsAndHashCode.Exclude
+    @ManyToOne
+    @JoinColumn(name = "order_id", nullable = false)
+    private Order order;
+    @ToString.Exclude
+    @EqualsAndHashCode.Exclude
+    @ManyToOne
```
Comment (BohdanTrue) [inline src/main/java/com/example/bookstore/model/OrderItem.java]: `@ManyToOne(fetch = FetchType.LAZY)`

File: src/main/resources/db/changelog/changes/18-create-orders-items-table.yaml [hunk 6]
```diff
@@ -0,0 +1,38 @@
+databaseChangeLog:
+  - changeSet:
+      id: create-orders-items-table
+      author: Taras Yurkovskyi
+      changes:
+        - createTable:
+            tableName: orders_items
```
Comment (BohdanTrue) [inline src/main/resources/db/changelog/changes/18-create-orders-items-table.yaml]: `tableName: order_items`

File: src/main/resources/db/changelog/changes/17-create-orders-table.yaml
Comment (BohdanTrue) [inline src/main/resources/db/changelog/changes/17-create-orders-table.yaml:1]: Where is is_deleted column?

File: src/main/resources/db/changelog/changes/18-create-orders-items-table.yaml
Comment (BohdanTrue) [inline src/main/resources/db/changelog/changes/18-create-orders-items-table.yaml:1]: same about is_deleted column

File: src/main/java/com/example/bookstore/model/Order.java
Comment (BohdanTrue) [inline src/main/java/com/example/bookstore/model/Order.java:1]: GJ, but don't forget about isDeleted field :)

File: src/main/java/com/example/bookstore/model/OrderItem.java
Comment (BohdanTrue) [inline src/main/java/com/example/bookstore/model/OrderItem.java:1]: same about isDeleted field

File: src/main/java/com/example/bookstore/service/impl/OrderServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+        return orderMapper.toDto(savedOrder);
+    }
+
+    @Override
+    public OrderResponseDto updateOrderStatus(Long id, UpdateOrderItemDto requestDto) {
+        Order order = orderRepository.findById(id).orElseThrow(
+                () -> new EntityNotFoundException("Cannot get order by id: " + id)
+        );
+        order.setStatus(requestDto.getStatus());
+
```
Comment (BohdanTrue) [inline src/main/java/com/example/bookstore/service/impl/OrderServiceImpl.java:72]: don't forget to save order with new status to db

File: src/main/java/com/example/bookstore/service/impl/OrderItemServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+                                                  Long orderId, Long orderItemId) {
+        Order order = orderRepository.findByIdAndUserId(orderId, userId).orElseThrow(
+                () -> new EntityNotFoundException("Cannot find order by id: " + orderId)
+        );
+        return orderItemMapper.toDto(order.getOrderItems().stream()
+                .filter(orderItem -> orderItem.getId().equals(orderItemId))
+                .findFirst()
+                .orElseThrow(
+                        () -> new EntityNotFoundException("Cannot find orderItem by id: "
+                                + orderItemId)));
```
Comment (BohdanTrue) [inline src/main/java/com/example/bookstore/service/impl/OrderItemServiceImpl.java]: GJ, but you can simplify it. Try to create method to find order item by query

**Review by BohdanTrue — CHANGES_REQUESTED** [review]
GJ, but there are some changes should be done!

File: src/main/resources/db/changelog/changes/17-create-orders-table.yaml
Comment (TarasYurkovskyi) [reply src/main/resources/db/changelog/changes/17-create-orders-table.yaml:1]: POST, PUT, and DELETE endpoints for OrderItem may not be necessary, as OrderItems are typically created, updated, or removed when a user interacts with their shopping cart or places an order.

File: src/main/java/com/example/bookstore/service/impl/OrderServiceImpl.java (hunk 7 above)
Comment (TarasYurkovskyi) [reply src/main/java/com/example/bookstore/service/impl/OrderServiceImpl.java:71]: return orderMapper.toDto(orderRepository.save(order)); 
the next line))

Comment (TarasYurkovskyi) [reply src/main/java/com/example/bookstore/service/impl/OrderServiceImpl.java:72]: return orderMapper.toDto(orderRepository.save(order));
the next line))

File: src/main/resources/db/changelog/changes/18-create-orders-items-table.yaml (hunk 6 above)
Comment (TarasYurkovskyi) [reply src/main/resources/db/changelog/changes/18-create-orders-items-table.yaml]: why order? in table there are not only items from one order

File: src/main/java/com/example/bookstore/controller/OrderController.java
Comment (d1sam) [inline src/main/java/com/example/bookstore/controller/OrderController.java:1]: Available for users with role ADMIN

- POST: /api/books/
- PUT: /api/books/{id}
- DELETE: /api/books/{id}
- POST: /api/categories
- PUT: /api/categories/{id}
- DELETE: /api/categories/{id}
- PATCH: /api/orders/{id}

File: src/main/java/com/example/bookstore/model/Order.java
Comment (d1sam) [reply src/main/java/com/example/bookstore/model/Order.java:1]: agree with @BohdanTrue :)

File: src/main/java/com/example/bookstore/model/OrderItem.java
Comment (d1sam) [reply src/main/java/com/example/bookstore/model/OrderItem.java:1]: agree with @BohdanTrue :)

File: src/main/java/com/example/bookstore/repository/OrderItemRepository.java [hunk 9]
```diff
@@ -0,0 +1,10 @@
+package com.example.bookstore.repository;
+
+import com.example.bookstore.model.OrderItem;
+import java.util.List;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface OrderItemRepository extends JpaRepository<OrderItem, Long> {
```
Comment (d1sam) [inline src/main/java/com/example/bookstore/repository/OrderItemRepository.java:12]:
```suggestion
@Repository
public interface OrderItemRepository extends JpaRepository<OrderItem, Long> {
```

File: src/main/resources/db/changelog/changes/18-create-orders-items-table.yaml (hunk 6 above)
Comment (d1sam) [reply src/main/resources/db/changelog/changes/18-create-orders-items-table.yaml]: agree with @BohdanTrue :)

File: src/main/java/com/example/bookstore/service/impl/OrderItemServiceImpl.java (hunk 8 above)
Comment (d1sam) [reply src/main/java/com/example/bookstore/service/impl/OrderItemServiceImpl.java]: yes, can be simplified with only one `repo` call

File: src/main/java/com/example/bookstore/service/impl/OrderServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+                        cartItem.getQuantity(), cartItem.getBook().getPrice()))
+                .collect(Collectors.toSet());
+
+        order.setOrderItems(orderItems);
+        cartItemRepository.deleteAll(shoppingCart.getCartItems());
+
+        return orderMapper.toDto(savedOrder);
+    }
+
+    @Override
```
Comment (d1sam) [inline src/main/java/com/example/bookstore/service/impl/OrderServiceImpl.java:65]:
```suggestion
    @Override
    @Transactional
```

**Review by d1sam — CHANGES_REQUESTED** [review]
good job, but there are some changes should be done!

**Review by vovasalyha — APPROVED** [review]

**Pull request merged**